contracts/Delay.sol

```diff
@@ -80,90 +136,198 @@ contract Delay is AragonApp, IForwarder {
     */
     function resumeExecution(uint256 _delayedScriptId) external auth(RESUME_EXECUTION_ROLE) {
-        require(_isExecutionPaused(_delayedScriptId), ERROR_CAN_NOT_RESUME);
         DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        require(delayedScript.disputableStatus == DisputableStatus.Active, ERROR_ACTION_NOT_ACTIVE);
 
-        uint64 timePaused = getTimestamp64().sub(delayedScript.pausedAt);
-        delayedScript.executionTime = delayedScript.executionTime.add(timePaused);
-        delayedScript.pausedAt = 0;
-
-        emit ExecutionResumed(_delayedScriptId);
+        _resumeExecution(delayedScript);
+        emit ExecutionResumed(_delayedScriptId, delayedScript.actionId);
     }
 
     /**
     * @notice Cancel script execution with ID `_delayedScriptId`
     * @param _delayedScriptId The ID of the script execution to cancel
     */
-    function cancelExecution(uint256 _delayedScriptId) external scriptExists(_delayedScriptId) auth(CANCEL_EXECUTION_ROLE) {
-        delete delayedScripts[_delayedScriptId];
+    function cancelExecution(uint256 _delayedScriptId) external nonReentrant auth(CANCEL_EXECUTION_ROLE) {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        require(_scriptExists(delayedScript), ERROR_NO_SCRIPT);
+        require(delayedScript.disputableStatus != DisputableStatus.Challenged, ERROR_ACTION_CHALLENGED);
+
+        _closeAgreementAction(delayedScript.actionId);
 
-        emit ExecutionCancelled(_delayedScriptId);
+        emit ExecutionCancelled(_delayedScriptId, delayedScript.actionId);
+        _deleteDelayedScript(_delayedScriptId);
     }
 
     /**
     * @notice Execute the script with ID `_delayedScriptId`
     * @param _delayedScriptId The ID of the script to execute
     */
-    function execute(uint256 _delayedScriptId) external {
-        require(canExecute(_delayedScriptId), ERROR_CAN_NOT_EXECUTE);
-
-        DelayedScript memory delayedScript = delayedScripts[_delayedScriptId];
-        delete delayedScripts[_delayedScriptId];
+    function execute(uint256 _delayedScriptId) external nonReentrant {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        require(_canExecute(delayedScript), ERROR_CAN_NOT_EXECUTE);
 
         runScript(delayedScript.evmCallScript, new bytes(0), new address[](0));
-        emit ExecutedScript(_delayedScriptId);
+
+        if (delayedScript.disputableStatus != DisputableStatus.Closed) {
+            _closeAgreementAction(delayedScript.actionId);
+        }
+
+        emit ExecutedScript(_delayedScriptId, delayedScript.actionId);
+        _deleteDelayedScript(_delayedScriptId);
+    }
+
+    /**
+    * @notice Close agreement action with ID `_delayedScriptId`
+    * @param _delayedScriptId The ID of the script to close the agreement action of
+    */
+    function closeAgreementAction(uint256 _delayedScriptId) external {
```

Comment:
Is there any rationale to have this function if `cancelExecution` is already provided?

Reply:
Cancel execution has been removed now.